### DISCUSSION TITLE
@app.on_event (startup and shutdown) - events don't work under daphne server

### DISCUSSION THREAD

**Discussion by starlingnic** [fastapi/fastapi#9444] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Request
from fastapi.middleware.cors import CORSMiddleware

from .config.db import init_db
from .config.ocpp_logging import logger
from .config.settings import settings
from .routers.auth_routs import auth_router
from .routers.geos_routs import address_router, region_router, country_router
from .routers.firms_routs import firms_router
from .routers.stations_routs import stations_router
from .routers.ocpp_routs import ocpp_router

print("\nsettings.PROJECT_NAME = ", settings.PROJECT_NAME)
logger.debug(f"===[00]=== = ")

logger.debug(f"===[start_application]=== = ")
app = FastAPI(
    title=settings.PROJECT_NAME,
    description=settings.PROJECT_DESCRIPTION,
    version=settings.PROJECT_VERSION,
    debug=True
)
app.include_router(country_router, prefix=settings.GEOS_PREFIX, tags=["Country"])
app.include_router(region_router, prefix=settings.GEOS_PREFIX, tags=["Region"])
app.include_router(address_router, prefix=settings.GEOS_PREFIX, tags=["Address"])
app.include_router(auth_router, prefix=settings.AUTH_PREFIX, tags=["Auth"])
app.include_router(firms_router, prefix=settings.FIRMS_PREFIX, tags=["Firm"])
app.include_router(stations_router, prefix=settings.STATIONS_PREFIX, tags=["Station"])
app.include_router(ocpp_router, prefix=settings.WS_PREFIX, tags=["Ocpp"])

origins = ["*"]
app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)


logger.debug(f"===[01]=== = ")


logger.debug(f"===[04]=== = ")
@app.on_event("startup")
async def startup_event():
    print(f"===[-05-]=== = ")
    logger.debug(f"===[05]=== = ")
    # await init_db(app)
    logger.debug(f"===[06]=== = ")

logger.debug(f"===[07]=== = ")


@app.on_event("shutdown")
async def shutdown_event():
    logger.debug(f"===[08]=== = ")
    pass

logger.debug(f"===[09]=== = ")

@app.get("/")
async def root():
    return {"message": "Hello!"}
```

### Description

FILE daphne.service:

[Unit]
Description=Daphne
After=network.target

[Service]
# Type=simple
User=geo
Group=www-data
WorkingDirectory=/srv/fapiprj
Environment="PATH=/srv/fapiprj-venv/bin"
ExecStart=/srv/fapiprj-venv/bin/daphne -u /srv/fapiprj/fapiprj.sock -b 0.0.0.0 -p 8001 -v 2 app.main:app
[Install]
WantedBy=multi-user.target

FILE app/config/ocpp_logging.py:

import logging
from functools import lru_cache

@lru_cache()
def get_logger():
    lgr = logging.getLogger()
    lgr.setLevel(logging.DEBUG)
    return lgr

logger = get_logger()

### Operating System

Linux

### Operating System Details

Debian 11

### FastAPI Version

0.95.1

### Python Version

3.92

### Additional Context

Log output:

апр 26 05:34:42 v3-dev systemd[1]: Stopping Daphne...
апр 26 05:34:42 v3-dev daphne[41996]: 2023-04-26 05:34:42,251 INFO     Received SIGTERM, shutting down.
апр 26 05:34:42 v3-dev daphne[41996]: 2023-04-26 05:34:42,291 INFO     Killed 0 pending application instances
апр 26 05:34:42 v3-dev daphne[41996]: 2023-04-26 05:34:42,324 INFO     (UNIX Port /srv/fapiprj/fapiprj.sock Closed)
апр 26 05:34:42 v3-dev daphne[41996]: 2023-04-26 05:34:42,327 INFO     (TCP Port 8001 Closed)
апр 26 05:34:42 v3-dev daphne[41996]: 2023-04-26 05:34:42,328 INFO     Stopping factory <daphne.http_protocol.HTTPFactory object at 0x7f52624da460>
апр 26 05:34:43 v3-dev daphne[41996]: settings.PROJECT_NAME =  testTEST
апр 26 05:34:43 v3-dev systemd[1]: daphne.service: Succeeded.
апр 26 05:34:43 v3-dev systemd[1]: Stopped Daphne.
апр 26 05:34:43 v3-dev systemd[1]: daphne.service: Consumed 26.237s CPU time.
апр 26 05:34:43 v3-dev systemd[1]: Started Daphne.
апр 26 05:34:43 v3-dev systemd[1]: Stopping Daphne...
апр 26 05:34:43 v3-dev systemd[1]: daphne.service: Succeeded.
апр 26 05:34:43 v3-dev systemd[1]: Stopped Daphne.
апр 26 05:34:43 v3-dev systemd[1]: Started Daphne.
апр 26 05:34:45 v3-dev daphne[42549]: 2023-04-26 05:34:45,814 DEBUG    ===[00]=== =
апр 26 05:34:45 v3-dev daphne[42549]: 2023-04-26 05:34:45,818 DEBUG    ===[start_application]=== =
апр 26 05:34:46 v3-dev daphne[42549]: 2023-04-26 05:34:46,325 DEBUG    ===[01]=== =
апр 26 05:34:46 v3-dev daphne[42549]: 2023-04-26 05:34:46,326 DEBUG    ===[04]=== =
апр 26 05:34:46 v3-dev daphne[42549]: 2023-04-26 05:34:46,326 DEBUG    ===[07]=== =
апр 26 05:34:46 v3-dev daphne[42549]: 2023-04-26 05:34:46,326 DEBUG    ===[09]=== =
апр 26 05:34:46 v3-dev daphne[42549]: 2023-04-26 05:34:46,332 INFO     Starting server at tcp:port=8001:interface=0.0.0.0, unix:/srv/fapiprj/fapiprj.sock
апр 26 05:34:46 v3-dev daphne[42549]: 2023-04-26 05:34:46,333 INFO     HTTP/2 support not enabled (install the http2 and tls Twisted extras)
апр 26 05:34:46 v3-dev daphne[42549]: 2023-04-26 05:34:46,333 INFO     Configuring endpoint tcp:port=8001:interface=0.0.0.0
апр 26 05:34:46 v3-dev daphne[42549]: 2023-04-26 05:34:46,336 INFO     HTTPFactory starting on 8001
апр 26 05:34:46 v3-dev daphne[42549]: 2023-04-26 05:34:46,336 INFO     Starting factory <daphne.http_protocol.HTTPFactory object at 0x7f8d40bf85e0>
апр 26 05:34:46 v3-dev daphne[42549]: 2023-04-26 05:34:46,337 INFO     Listening on TCP address 0.0.0.0:8001
апр 26 05:34:46 v3-dev daphne[42549]: 2023-04-26 05:34:46,337 INFO     Configuring endpoint unix:/srv/fapiprj/fapiprj.sock
апр 26 05:34:46 v3-dev daphne[42549]: 2023-04-26 05:34:46,337 INFO     HTTPFactory starting on '/srv/fapiprj/fapiprj.sock'

**Comment by nymous** [comment, marked as answer]
Hello!
Unfortunately it seems that Daphne doesn't support ASGI lifespan, so the events will not be fired (see https://github.com/django/daphne/issues/264).
You could migrate to a server that supports them, like [Uvicorn](https://www.uvicorn.org/), or maybe try a workaround with something like [asgi-lifespan](https://pypi.org/project/asgi-lifespan/) (but I haven't tested that at all, just throwing the idea).
